std/algorithm/iteration.d

```diff
@@ -902,67 +913,100 @@ template each(alias pred = "a")
             size_t i = 0;
             while (!r.empty)
             {
-                cast(void) binaryFun!BinaryArgs(i, r.front);
+                static if (!is(typeof(binaryFun!BinaryArgs(i, r.front)) ==
+                        Flag!"each"))
+                    cast(void) binaryFun!BinaryArgs(i, r.front);
+                else
+                    if (binaryFun!BinaryArgs(i, r.front) == No.each)
+                        return No.each;
                 r.popFront();
                 i++;
             }
         }
+        return Yes.each;
     }
 
-    void each(Iterable)(auto ref Iterable r)
-        if (isForeachIterable!Iterable)
+    Flag!"each" each(Iterable)(auto ref Iterable r)
+    if (isForeachIterable!Iterable)
     {
         debug(each) pragma(msg, "Using foreach for ", Iterable.stringof);
         static if (isForeachUnaryIterable!Iterable)
         {
             foreach (ref e; r)
-                cast(void) unaryFun!pred(e);
+                static if (!is(typeof(unaryFun!fun(e)) == Flag!"each"))
+                    cast(void) unaryFun!fun(e);
+                else
+                    if (unaryFun!fun(e) == No.each) return No.each;
         }
         else // if (isForeachBinaryIterable!Iterable)
         {
             foreach (ref i, ref e; r)
-                cast(void) binaryFun!BinaryArgs(i, e);
+                static if (!is(typeof(binaryFun!BinaryArgs(i, e)) == Flag!"each"))
+                    cast(void) binaryFun!BinaryArgs(i, e);
+                else
+                    if (binaryFun!BinaryArgs(i, e) == No.each) return No.each;
         }
+        return Yes.each;
     }
 
     // opApply with >2 parameters. count the delegate args.
     // only works if it is not templated (otherwise we cannot count the args)
-    void each(Iterable)(auto ref Iterable r)
+    Flag!"each" each(Iterable)(auto ref Iterable r)
         if (!isRangeIterable!Iterable && !isForeachIterable!Iterable &&
             __traits(compiles, Parameters!(Parameters!(r.opApply))))
     {
-        auto dg(Parameters!(Parameters!(r.opApply)) params) {
-            pred(params);
-            return 0; // tells opApply to continue iteration
+        auto result = Yes.each;
+        auto dg(Parameters!(Parameters!(r.opApply)) params)
+        {
+            static if (!is(typeof(binaryFun!BinaryArgs(i, e)) == Flag!"each"))
```

Comment:
`is(typeof(binaryFun!BinaryArgs(i, e)) == Flag!"each")` appears repeatedly. Factor it out and use it to initialize a manifest constant.